bundles/org.eclipse.core.commands/src/org/eclipse/core/commands/INamedHandler.java

```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Vector Informatik GmbH and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Niklas Kellner - initial API
+ *******************************************************************************/
+package org.eclipse.core.commands;
+
+/**
+ * Extends the IHandler interface to provide a way for an handler to define a
+ * Name. A use case is the Undo/Redo command where the handler specifies which
+ * action will be undone.
+ *
+ * @since 3.12
+ *
+ */
+public interface INamedHandler extends IHandler {
+
+	/**
+	 * Called by the framework to allow the handler to specify a name.
+	 *
+	 * @return name of the Handler
+	 */
+	String getHandlerName();
```

Comment:
Sorry if I already asked this, but was it considered to add this as a default method in IHandler directly?

Reply:
When I implemented this my intention was to not add it to the IHandler interface directly as this is a somewhat niche use case and not supported everywhere where the IHandler is used.

What I think is open for discussion is that if this feature is something that should be generally supported, then also a proper E4 support is needed with an annotation (Named?) that provides this name.

But this should probably be handled in a separate PR.

Reply:
> But this should probably be handled in a separate PR.

As this is API, we unfortunately won't be able to roll back if we add it.
As I'm looking more to #1314, I'm wondering why do we need labels on handlers here. Don't we already have a name on the command? Or is it that some operations have handlers without commands? (sorry for joining the discussion late and probably bugging with questions that were already considered)

Reply:
The handler can override the label from the command in some special cases, only in e3.

In E4 this is not possible, and we switched the Undo Redo handler to an E4 based handler bridge. (iirc)

There are a few cases where this is used and one case is the undo/redo. So the label basically says "Undo Typing" and not only "Undo" which is the name of the command.

Reply:
OK, got it. And indeed, if we want some dynamic label, only the handler can provide it.
However, I don't think it's so much of a "niche use-case" or at least that it's a use-case that we want to keep as "niche". I find it interesting that handlers enrich the action label to make it more precise. It's a good practice that should be kind of recommended.
So I think it's better to make it a new default method to encourage and make it easier for implementors in enriching the label.

Reply:
You are right that it provides a benefit, but I think my current implementation is not ready enough for a general rollout.

As you approved the PR I'm a bit confused now, should I move it now to the IHandler interface? Or should this stay in a separate interface as it is optional? (for now?)

Reply:
> Or should this stay in a separate interface as it is optional?

default methods are optional to implement as well.

We don't strongly care about implementation here, we care more about APIs (which we can't easily change later).
I would still recommend that you make it a new default method in IHandler if possible, because it actually seems to provide more value than the current extension interface.
While you do that, I think it'd be better to remame the method to `getHandlerLabel()` because the name can be interpreted as an "id" which is not the case here. (Thinking more about it, it could be just `getLabel()` but there would be too high risks of conflicts so let's stick to `getHandlerLabel()`, and this makes me think that we might instead of a new API it might be possible to try relying on ILabelProvider and adapters, but this would be more complicated for no strong value and also less encouraging than a new default method in API, so let's stick to a default method).

Reply:
You should add an API filter.  Only the minor version should be increased.  All hell will break loose if you increment the major version.

Reply:
@merks Thanks, I already noticed this. I will figure out how this works with an API filter

Reply:
There should be a quick fix that does the work for you.  👍